### PR TITLE
Fix #2936 - boost::bind -> std::bind and using std::placeholders

### LINKED ISSUE #2936: error: ‘_1’ was not declared in this scope
Bug Report
==========

Environment
-----------

* **FreeOrion Version**: b7a2164a8d7b6889e96488ed1ebdb065fa1647a0
* **Operating System**: Gentoo GNU/Linux

Description
-----------

```
‰ make                                                                             
[  0%] Building CXX object GG/CMakeFiles/GiGi.dir/src/dialogs/ColorDlg.cpp.o
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp: In member function ‘virtual void GG::ColorDlg::CompleteConstruction()’:
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:622:56: error: ‘_1’ was not declared in this scope
  622 |         boost::bind(&ColorDlg::RedSliderChanged, this, _1, _2, _3));
      |                                                        ^~
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:622:56: note: suggested alternatives:
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:29:16: note:   ‘mpl_::_1’
   29 | typedef arg<1> _1;
      |                ^~
In file included from /usr/include/boost/algorithm/string/detail/classification.hpp:16,
                 from /usr/include/boost/algorithm/string/classification.hpp:18,
                 from /usr/include/boost/algorithm/string/trim.hpp:23,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/functional:211:34: note:   ‘std::placeholders::_1’
  211 |     extern const _Placeholder<1> _1;
      |                                  ^~
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:29:16: note:   ‘mpl_::_1’
   29 | typedef arg<1> _1;
      |                ^~
In file included from /usr/include/boost/bind/bind.hpp:2356,
                 from /usr/include/boost/signals2/slot.hpp:15,
                 from /usr/include/boost/signals2/connection.hpp:24,
                 from /usr/include/boost/signals2/signal.hpp:22,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/bind/placeholders.hpp:46:38: note:   ‘boost::placeholders::_1’
   46 | BOOST_STATIC_CONSTEXPR boost::arg<1> _1;
      |                                      ^~
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:29:16: note:   ‘mpl_::_1’
   29 | typedef arg<1> _1;
      |                ^~
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:622:60: error: ‘_2’ was not declared in this scope
  622 |         boost::bind(&ColorDlg::RedSliderChanged, this, _1, _2, _3));
      |                                                            ^~
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:622:60: note: suggested alternatives:
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:42:16: note:   ‘mpl_::_2’
   42 | typedef arg<2> _2;
      |                ^~
In file included from /usr/include/boost/algorithm/string/detail/classification.hpp:16,
                 from /usr/include/boost/algorithm/string/classification.hpp:18,
                 from /usr/include/boost/algorithm/string/trim.hpp:23,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/functional:212:34: note:   ‘std::placeholders::_2’
  212 |     extern const _Placeholder<2> _2;
      |                                  ^~
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:42:16: note:   ‘mpl_::_2’
   42 | typedef arg<2> _2;
      |                ^~
In file included from /usr/include/boost/bind/bind.hpp:2356,
                 from /usr/include/boost/signals2/slot.hpp:15,
                 from /usr/include/boost/signals2/connection.hpp:24,
                 from /usr/include/boost/signals2/signal.hpp:22,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/bind/placeholders.hpp:47:38: note:   ‘boost::placeholders::_2’
   47 | BOOST_STATIC_CONSTEXPR boost::arg<2> _2;
      |                                      ^~
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:42:16: note:   ‘mpl_::_2’
   42 | typedef arg<2> _2;
      |                ^~
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:622:64: error: ‘_3’ was not declared in this scope
  622 |         boost::bind(&ColorDlg::RedSliderChanged, this, _1, _2, _3));
      |                                                                ^~
/home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:622:64: note: suggested alternatives:
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:55:16: note:   ‘mpl_::_3’
   55 | typedef arg<3> _3;
      |                ^~
In file included from /usr/include/boost/algorithm/string/detail/classification.hpp:16,
                 from /usr/include/boost/algorithm/string/classification.hpp:18,
                 from /usr/include/boost/algorithm/string/trim.hpp:23,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/functional:213:34: note:   ‘std::placeholders::_3’
  213 |     extern const _Placeholder<3> _3;
      |                                  ^~
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:55:16: note:   ‘mpl_::_3’
   55 | typedef arg<3> _3;
      |                ^~
In file included from /usr/include/boost/bind/bind.hpp:2356,
                 from /usr/include/boost/signals2/slot.hpp:15,
                 from /usr/include/boost/signals2/connection.hpp:24,
                 from /usr/include/boost/signals2/signal.hpp:22,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/bind/placeholders.hpp:48:38: note:   ‘boost::placeholders::_3’
   48 | BOOST_STATIC_CONSTEXPR boost::arg<3> _3;
      |                                      ^~
In file included from /usr/include/boost/mpl/aux_/include_preprocessed.hpp:37,
                 from /usr/include/boost/mpl/placeholders.hpp:43,
                 from /usr/include/boost/iterator/iterator_categories.hpp:16,
                 from /usr/include/boost/iterator/iterator_facade.hpp:13,
                 from /usr/include/boost/range/iterator_range_core.hpp:27,
                 from /usr/include/boost/range/iterator_range.hpp:13,
                 from /usr/include/boost/range/as_literal.hpp:22,
                 from /usr/include/boost/algorithm/string/trim.hpp:19,
                 from /home/l29ah/projects/freeorion/GG/GG/Enum.h:40,
                 from /home/l29ah/projects/freeorion/GG/GG/Base.h:38,
                 from /home/l29ah/projects/freeorion/GG/GG/Wnd.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/Control.h:32,
                 from /home/l29ah/projects/freeorion/GG/GG/TextControl.h:33,
                 from /home/l29ah/projects/freeorion/GG/GG/Button.h:35,
                 from /home/l29ah/projects/freeorion/GG/GG/dialogs/ColorDlg.h:32,
                 from /home/l29ah/projects/freeorion/GG/src/dialogs/ColorDlg.cpp:25:
/usr/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:55:16: note:   ‘mpl_::_3’
   55 | typedef arg<3> _3;
      |                ^~
make[2]: *** [GG/CMakeFiles/GiGi.dir/build.make:83: GG/CMakeFiles/GiGi.dir/src/dialogs/ColorDlg.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:774: GG/CMakeFiles/GiGi.dir/all] Error 2
make: *** [Makefile:172: all] Error 2
```

### PR DESCRIPTION
**Pull request by geoffthemedio**
Attempting to fix https://github.com/freeorion/freeorion/issues/2936

### CODE REVIEW

File: GG/src/Font.cpp [hunk 1]
```diff
@@ -506,6 +506,8 @@ namespace {
             xpr::mark_tag whitespace_tag(4);
             xpr::mark_tag text_tag(5);
 
+            using std::placeholders::_1;
```
Comment (adrianbroher) [inline GG/src/Font.cpp:509]: Not needed, as you are addressing the placeholders with a qualified name.  So either drop this line or remove the namespace qualifier at their place of use.

File: GG/src/GUI.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@
 #include <iostream>
 #include <fstream>
 #include <list>
-
+#include <functional>
```
Comment (adrianbroher) [inline GG/src/GUI.cpp]: Please sort the headers alphabetically.

File: GG/src/dialogs/ColorDlg.cpp [hunk 3]
```diff
@@ -614,34 +614,38 @@ void ColorDlg::CompleteConstruction()
     master_layout->Add(m_sliders_ok_cancel_layout, 0, 1, 3, 1);
     SetLayout(master_layout);
 
+    using std::placeholders::_1;
```
Comment (adrianbroher) [inline GG/src/dialogs/ColorDlg.cpp:617]: You can also use

`using namespace std::placeholder;`

to make all placeholders visible in one statement.

**Review by adrianbroher — COMMENTED** [review]
I doubt that replacing this within GiGi only fix the problem.  This probably has do be done in the whole code base.

File: GG/src/dialogs/ColorDlg.cpp (hunk 3 above)
Comment (geoffthemedio) [reply GG/src/dialogs/ColorDlg.cpp:617]: Not in general; it often results in ambiguity with the `boost::placeholders::_1`

**Comment by geoffthemedio** [comment]
> I doubt that replacing this within GiGi only fix the problem. This probably has do be done in the whole code base.

Hence this being a draft. WIP.

File: universe/Pathfinder.cpp [hunk 4]
```diff
@@ -1332,13 +1337,15 @@ bool Pathfinder::PathfinderImpl::WithinJumpsOfOthers(
...
     // Examine the cache to see if \p system_id is within \p jumps of \p others
     bool within_jumps(false);
     distance_matrix_cache<distance_matrix_storage<short>> cache(m_system_jumps);
     cache.examine_row(system_index,
-                      boost::bind(&Pathfinder::PathfinderImpl::HandleCacheMiss, this, _1, _2),
-                      boost::bind(&Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit, this,
-                                  boost::ref(within_jumps), jumps, others, _1, _2));
+                      std::bind(&Pathfinder::PathfinderImpl::HandleCacheMiss, this, ph::_1, ph::_2),
+                      std::bind(&Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit, this,
+                                boost::ref(within_jumps), jumps, others, ph::_1, ph::_2));
```
Comment (adrianbroher) [inline universe/Pathfinder.cpp]: This needs to be converted to `std::ref`.

Comment (geoffthemedio) [reply universe/Pathfinder.cpp]: What specifically, and is that related to the placeholders issue this is addressing?

File: client/ClientNetworking.cpp [hunk 5]
```diff
@@ -118,6 +118,7 @@ namespace {
                                 this,
                                 boost::asio::placeholders::error,
                                 boost::asio::placeholders::bytes_transferred));
+
```
Comment (adrianbroher) [inline client/ClientNetworking.cpp:121]: Is this whitespace intentional or just slipped in?

File: UI/BuildDesignatorWnd.cpp [hunk 6]
```diff
@@ -30,6 +30,7 @@
 
 #include <iterator>
 
+using namespace std::placeholders;
```
Comment (adrianbroher) [inline UI/BuildDesignatorWnd.cpp]: Another dangling namespace using statement that should not be required.

File: client/ClientNetworking.cpp (hunk 5 above)
Comment (geoffthemedio) [reply client/ClientNetworking.cpp:121]: it was added intentionally between when I replaced all the `boost::bind` and when I reverted most of the networking ones that use `asio::placeholders` such as the above.

File: universe/Pathfinder.cpp (hunk 4 above)
Comment (adrianbroher) [reply universe/Pathfinder.cpp]: The `within_jumps` parameter of that function should look like:

```
 std::ref(within_jumps)
```

**Comment by adrianbroher** [comment]
Except the remarks looks good to me.

File: universe/Pathfinder.cpp (hunk 4 above)
Comment (adrianbroher) [reply universe/Pathfinder.cpp]: > and is that related to the placeholders issue this is addressing?

I'm not sure if you can mix and match `boost::ref` with `std::bind` without problems.

**Comment by l29ah** [comment]
Tried to build it, got a few of these:
```
In file included from /home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:3:
/home/l29ah/projects/freeorion/parse/ParseImpl.h:29:40: error: ‘emplace_back_1_’ has not been declared
   29 |     BOOST_PHOENIX_ADAPT_FUNCTION(void, emplace_back_1_, emplace_back_1, 2)
      |                                        ^~~~~~~~~~~~~~~
/home/l29ah/projects/freeorion/parse/ParseImpl.h:29:57: error: ‘emplace_back_1’ is not a type
   29 |     BOOST_PHOENIX_ADAPT_FUNCTION(void, emplace_back_1_, emplace_back_1, 2)
      |                                                         ^~~~~~~~~~~~~~
/home/l29ah/projects/freeorion/parse/ParseImpl.h:29:73: error: expected identifier before numeric constant
   29 |     BOOST_PHOENIX_ADAPT_FUNCTION(void, emplace_back_1_, emplace_back_1, 2)
      |                                                                         ^
/home/l29ah/projects/freeorion/parse/ParseImpl.h:29:73: error: expected ‘,’ or ‘...’ before numeric constant
/home/l29ah/projects/freeorion/parse/ParseImpl.h:32:5: error: expected constructor, destructor, or type conversion before ‘struct’
   32 |     struct is_unique {
      |     ^~~~~~
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:26:51: error: ‘is_unique’ is not a member of ‘parse::detail’
   26 |     const boost::phoenix::function<parse::detail::is_unique> is_unique_;
      |                                                   ^~~~~~~~~
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:26:51: error: ‘is_unique’ is not a member of ‘parse::detail’
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:26:60: error: template argument 1 is invalid
   26 |     const boost::phoenix::function<parse::detail::is_unique> is_unique_;
      |                                                            ^
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:26:62: error: uninitialized const ‘{anonymous}::is_unique_’ [-fpermissive]
   26 |     const boost::phoenix::function<parse::detail::is_unique> is_unique_;
      |                                                              ^~~~~~~~~~
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp: In constructor ‘{anonymous}::grammar::grammar(const parse::lexer&, const string&, const text_iterator&, const text_iterator&)’:
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:85:49: error: ‘{anonymous}::is_unique_’ cannot be used as a function
   85 |                 [ _pass = is_unique_(_r1, _1, _2),
      |                                                 ^
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp: At global scope:
/home/l29ah/projects/freeorion/parse/BuildingsParser.cpp:28:10: warning: ‘void {anonymous}::insert_building(std::map<std::__cxx11::basic_string<char>, std::unique_ptr<BuildingType> >&, const string&, const string&, const parse::detail::MovableEnvelope<CommonParams>&, CaptureResult&, const string&, bool&)’ defined but not used [-Wunused-function]
   28 |     void insert_building(std::map<std::string, std::unique_ptr<BuildingType>>& building_types,
      |          ^~~~~~~~~~~~~~~
```

**Comment by adrianbroher** [comment]
Those lines can be dropped, this is dead code:

https://github.com/freeorion/freeorion/blob/master/parse/ParseImpl.h#L24-L30

The last usage was removed with 2ff93a4d278aa402c7fa88ff886f1a5ec2245048 2.5 years ago.

**Comment by l29ah** [comment]
Okay, what about this 400kB worth of an error message?
[error.log](https://github.com/freeorion/freeorion/files/4605955/error.log)

**Comment by geoffthemedio** [comment]
> Okay, what about this 400kB worth of an error message?

What Boost version were you using before?

**Comment by l29ah** [comment]
Must be 1.72.0.

**Comment by geoffthemedio** [comment]
> Okay, what about this 400kB worth of an error message?
> [error.log](https://github.com/freeorion/freeorion/files/4605955/error.log)

That appears to be 22 error messages. Is that all of them? They are all in EnumParser.cpp which suggests something local to that file might be the issue, but if there are others in other files, then that's more complicated.

If it is just in EnumParser.cpp then a bit of a shot in the dark is to try:
````
 parse/EnumParser.cpp | 26 ++++++++++++++++++++------
 1 file changed, 20 insertions(+), 6 deletions(-)

diff --git a/parse/EnumParser.cpp b/parse/EnumParser.cpp
index 784c7ebe7..4d47fb36c 100644
--- a/parse/EnumParser.cpp
+++ b/parse/EnumParser.cpp
@@ -5,18 +5,14 @@
 #include "../universe/UnlockableItem.h"
 
 
-namespace qi = boost::spirit::qi;
-
 #define DEBUG_PARSERS 0
 
-namespace {
-    qi::_val_type _val;
-}
-
 namespace parse {
     empire_affiliation_enum_grammar::empire_affiliation_enum_grammar(const parse::lexer& tok) :
         empire_affiliation_enum_grammar::base_type(rule, "empire_affiliation_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.TheEmpire_  [ _val = AFFIL_SELF ]
             |   tok.EnemyOf_    [ _val = AFFIL_ENEMY ]
@@ -32,6 +28,8 @@ namespace parse {
     unlockable_item_enum_grammar::unlockable_item_enum_grammar(const parse::lexer& tok) :
         unlockable_item_enum_grammar::base_type(rule, "unlockable_item_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.Building_       [ _val = UIT_BUILDING ]
             |   tok.ShipPart_       [ _val = UIT_SHIP_PART ]
@@ -44,6 +42,8 @@ namespace parse {
     ship_slot_enum_grammar::ship_slot_enum_grammar(const parse::lexer& tok) :
         ship_slot_enum_grammar::base_type(rule, "ship_slot_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.External_   [ _val = SL_EXTERNAL ]
             |   tok.Internal_   [ _val = SL_INTERNAL ]
@@ -54,6 +54,8 @@ namespace parse {
     ship_part_class_enum_grammar::ship_part_class_enum_grammar(const parse::lexer& tok) :
         ship_part_class_enum_grammar::base_type(rule, "ship_part_class_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.ShortRange_         [ _val = PC_DIRECT_WEAPON ]
             |   tok.FighterBay_         [ _val = PC_FIGHTER_BAY ]
@@ -78,6 +80,8 @@ namespace parse {
     capture_result_enum_grammar::capture_result_enum_grammar(const parse::lexer& tok) :
         capture_result_enum_grammar::base_type(rule, "capture_result_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.Capture_            [ _val = CR_CAPTURE ]
             |   tok.Retain_             [ _val = CR_RETAIN ]
@@ -88,6 +92,8 @@ namespace parse {
     statistic_enum_grammar::statistic_enum_grammar(const parse::lexer& tok) :
         statistic_enum_grammar::base_type(rule, "statistic_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.CountUnique_    [ _val = ValueRef::UNIQUE_COUNT ]
             |   tok.Count_          [ _val = ValueRef::COUNT ]
@@ -107,6 +113,8 @@ namespace parse {
     non_ship_part_meter_enum_grammar::non_ship_part_meter_enum_grammar(const parse::lexer& tok) :
         non_ship_part_meter_enum_grammar::base_type(rule, "non_ship_part_meter_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.TargetConstruction_     [ _val = METER_TARGET_CONSTRUCTION ]
             |   tok.TargetIndustry_         [ _val = METER_TARGET_INDUSTRY ]
@@ -149,6 +157,8 @@ namespace parse {
     ship_part_meter_enum_grammar::ship_part_meter_enum_grammar(const parse::lexer& tok) :
         ship_part_meter_enum_grammar::base_type(rule, "ship_part_meter_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.MaxCapacity_            [ _val = METER_MAX_CAPACITY ]
             |   tok.MaxDamage_              [ _val = METER_MAX_CAPACITY ]
@@ -162,6 +172,8 @@ namespace parse {
     set_non_ship_part_meter_enum_grammar::set_non_ship_part_meter_enum_grammar(const parse::lexer& tok) :
         set_non_ship_part_meter_enum_grammar::base_type(rule, "set_non_ship_part_meter_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.SetTargetConstruction_  [ _val = METER_TARGET_CONSTRUCTION ]
             |   tok.SetTargetIndustry_      [ _val = METER_TARGET_INDUSTRY ]
@@ -205,6 +217,8 @@ namespace parse {
     set_ship_part_meter_enum_grammar::set_ship_part_meter_enum_grammar(const parse::lexer& tok) :
         set_ship_part_meter_enum_grammar::base_type(rule, "set_ship_part_meter_enum_grammar")
     {
+        boost::spirit::qi::_val_type _val;
+
         rule
             =   tok.SetMaxCapacity_         [ _val = METER_MAX_CAPACITY ]
             |   tok.SetMaxDamage_           [ _val = METER_MAX_CAPACITY ]

````

**Comment by l29ah** [comment]
Not much different:
[build.log](https://github.com/freeorion/freeorion/files/4606161/build.log)

**Pull request merged**